OSDOCS-14942: corrections to ingress controller docs MicroShift (#94817)
openshift/openshift-docs: modules/microshift-ingress-controller-config.adoc

* You installed the OpenShift CLI (`oc`).
* You have root access to the cluster.
* Your cluster uses the OVN-Kubernetes Container Network Interface (CNI) plugin.

.Procedure

. Apply ingress control settings in one of the two following ways:

.. Update the {microshift-short} `config.yaml` configuration file by making a copy of the provided `config.yaml.default` file in the `/etc/microshift/` directory, naming it `config.yaml` and keeping it in the source directory.
+
[IMPORTANT]
====

> After you create it, the `config.yaml` file takes precedence over built-in settings. The configuration file is read every time the {microshift-short} service starts.

Reviewer: The _it_ here is a bit vague. Is this right:
```suggestion
After you create the `config.yaml` file, it takes precedence over built-in settings. The configuration file is read every time the {microshift-short} service starts.
```

* In-use generated and user-specified certificates are automatically integrated with the {microshift-short} built-in OAuth server.

|`clientTLS`
|Authenticates client access to the cluster and services. As a result, mutual TLS authentication is enabled. If not set, then client TLS is not enabled. You must set the `spec.clientTLS.clientCertificatePolicy` and `spec.clientTLS.ClientCA` parameters to use client TLS.
//Does this mean we can only use ASP if we are also using the other two? it is dependent?

|`clientTLS.AllowedSubjectPatterns`
|Optional subfield which specifies a list of regular expressions that are matched against the distinguished name on a valid client certificate to filter requests. Use this parameter to cause the ingress controller to reject certificates based on the distinguished name. The Perl Compatible Regular Expressions (PCRE) syntax is required.

[IMPORTANT]
====
When configured, this field must contain a valid expression or the {microshift-short} service fails. At least one pattern must match a client certificate's distinguished name; otherwise, the ingress controller rejects the certificate and denies the connection.
====

|`clientTLS.ClientCA`

> |Specifies a required config map that is in the `openshift-ingress` namespace. Required to enable client TLS. The config map must contain a certificate authority (CA) certificate bundle.

Reviewer: Probably redundant.
```suggestion
|Specifies a required config map that is in the `openshift-ingress` namespace. Required to enable client TLS. The config map must contain a certificate authority (CA) bundle.
```

|`forwardedHeaderPolicy`
|Specifies when and how the ingress controller sets the `Forwarded`, `X-Forwarded-For`, `X-Forwarded-Host`, `X-Forwarded-Port`, `X-Forwarded-Proto`, and `X-Forwarded-Proto-Version` HTTP headers. The following values are valid:

* `Append` is the default value. Preserves any existing headers by specifying that the ingress controller appends them.
* `Replace` removes any existing headers by specifying that the ingress controller sets the headers.
* `IfNone` sets the headers set by specifying that the ingress controller sets the headers if they are not already set.
* `Never` preserves any existing headers by specifying that the ingress controller never sets the headers.

|`httpCompression`
|Defines the policy for HTTP traffic compression.

> |`httpCompression.mimeTypes`
> |Defines a list of MIME types to which compression should be applied.
> 
> * For example, `text/css; charset=utf-8`, `text/html`, `text/*`, `image/svg+xml`, `application/octet-stream`, `X-custom/customsub`, using the format pattern, `type/subtype; [;attribute=value]`.
> * Valid `types` are: application, image, message, multipart, text, video, or a custom type prefaced by `X-`. To see the full notation for MIME types and subtypes, see link:https://datatracker.ietf.org/doc/html/rfc1341#page-7[RFC1341] (IETF Datatracker documentation).

Reviewer: ```suggestion
|`httpCompression.mimeTypes`
|Defines a list of MIME types to which compression should be applied, in the `type/subtype; [;attribute=value]` format.

* For example, `text/css; charset=utf-8`, `text/html`, `text/*`, `image/svg+xml`, `application/octet-stream`, `X-custom/customsub`.
* Valid `types` are: application, image, message, multipart, text, video, or a custom type prefaced by `X-`. To see the full notation for MIME types and subtypes, see link:https://datatracker.ietf.org/doc/html/rfc1341#page-7[RFC1341] (IETF Datatracker documentation).
```

|`httpEmptyRequestsPolicy`

> |Describes how HTTP connections are handled if the connection times out before a request is received. Allowed values for this field are `Respond` and `Ignore`. The default value is `Respond`. Empty requests typically come from load-balancer health probes or preconnects and can often be safely ignored. However, these requests can also be caused by network errors and port scans, so setting this field to `Ignore` can impede detection or diagnosis of network problems and detecting intrusion attempts.

Reviewer: Per ISG: [Do not use (so) by itself as a conjunction](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#s)

```suggestion
|Describes how HTTP connections are handled if the connection times out before a request is received. Allowed values for this field are `Respond` and `Ignore`. The default value is `Respond`. Empty requests typically come from load-balancer health probes or preconnects and can often be safely ignored. However, these requests can also be caused by network errors and port scans. Ass such setting this field to `Ignore` can impede detection or diagnosis of network problems and detecting intrusion attempts.
```

* When the policy is set to `Respond`, the ingress controller sends an HTTP `400` or `408` response, logs the connection if access logging is enabled, and counts the connection in the appropriate metrics.

> * When the policy is set to `Ignore`, the `http-ignore-probes` parameter is added in the `HAproxy` configuration. The ingress controller closes the connection without sending a response, then logs the connection or incrementing metrics.

Author: Specialty networking term that describes multiple types of networking metrics that are interpreted within a time series: _"Incrementing metrics" typically refers to tracking metrics that show a cumulative count or rate of events over time. These metrics are often used to analyze trends, identify performance degradation, and troubleshoot issues._

|`logEmptyRequests`

> |Specifies connections for which no request is received and logged. `Log` and `Ignore` are valid values. Empty requests typically come from load-balancer health probes or preconnects and can often be safely ignored. However, these requests can also be caused by network errors and port scans, so setting this field to `Ignore` can impede detection or diagnosis of network problems and detecting intrusion attempts. The default value is `Log`.

Reviewer: ```suggestion
|Specifies connections for which no request is received and logged. `Log` and `Ignore` are valid values. Empty requests typically come from load-balancer health probes or preconnects and can often be safely ignored. However, these requests can also be caused by network errors and port scans. As such, setting this field to `Ignore` can impede detection or diagnosis of network problems and detecting intrusion attempts. The default value is `Log`.
```

|`tlsSecurityProfile`
|`tlsSecurityProfile` specifies settings for TLS connections for ingress controllers. If not set, the default value is based on the `apiservers.config.openshift.io/cluster` resource. The TLS `1.0` version of an `Old` or `Custom` profile is automatically converted to `1.1` by the ingress controller. `Intermediate` is the default setting.

* The minimum TLS version for ingress controllers is `1.1`. The maximum TLS version is `1.3`.

[NOTE]
====
Ciphers and the minimum TLS version of the configured security profile are reflected in the `TLSProfile` status. Profiles profiles are intent-based and change over time as new ciphers are developed and existing ciphers are found to be insecure. The usable list can be reduced depending on which ciphers are available to a specific process.
====

|`tlsSecurityProfile.custom`
|User-defined TLS security profile. Configure this and the related parameters with extreme caution.

|`tlsSecurityProfile.custom.ciphers`
|Used to specify the cipher algorithms that are negotiated during the TLS handshake. Operators might remove entries their operands do not support.

|`tlsSecurityProfile.custom.minTLSVersion`

> |Used to specify the minimal version of the TLS protocol that is negotiated during the TLS handshake. For example, to use TLS versions 1.1, 1.2 and 1.3, set the value to `VersionTLS11`. The highest valid value for `minTLSVersion` is `VersionTLS12`.

Reviewer: ??
```suggestion
|Specifies the minimal version of the TLS protocol that is negotiated during the TLS handshake. For example, to use TLS versions 1.1, 1.2 and 1.3, set the value to `VersionTLS11`. The highest valid value for `minTLSVersion` is `VersionTLS12`.
```

[IMPORTANT]
====
Setting this field not recommended because `headerBufferMaxRewriteBytes` parameter values that are too small can break the ingress controller. Conversely, values for `headerBufferMaxRewriteBytes` that are too large could cause the ingress controller to use significantly more memory than necessary.
====

|`tuningOptions.headerBufferMaxRewriteBytes`
|Specifies how much memory should be reserved, in bytes, from `headerBufferBytes` for HTTP header rewriting and appending for ingress controller connection sessions. The minimum value for `headerBufferMaxRewriteBytes` is `4096`. `headerBufferBytes` must be greater than the `headerBufferMaxRewriteBytes` value for incoming HTTP requests. If not set, the default value is `8192` bytes.

[IMPORTANT]
====
Setting this field is not recommended because `headerBufferMaxRewriteBytes` values that are too small can break the ingress controller and `headerBufferMaxRewriteBytes` that are too large could cause the ingress controller to use significantly more memory than necessary.
====

|`tuningOptions.healthCheckInterval`

> |Specifies how long the router waits between health checks. The default is `5s`.

Reviewer: Do we need to state the allows units?
```suggestion
|Specifies how long the router waits between health checks, in seconds. The default is `5s`.
```

|`tuningOptions.maxConnections`

> |Specifies the maximum number of simultaneous connections that can be established per `HAProxy` process. Increasing this value allows each ingress controller pod to handle more connections at the cost of additional system resources. Permitted values are `0`, `-1`, any value within the range `2000` and `2000000`, or the field can be left empty.

Reviewer: Maybe a bit more clear?

```suggestion
|Specifies the maximum number of simultaneous connections that can be established for each `HAProxy` process. Increasing this value allows each ingress controller pod to handle more connections at the cost of additional system resources. Permitted values are `0`, `-1`, any value within the range `2000` and `2000000`, or the field can be left empty.
```

* If this field is left empty or has the value `0`, the ingress controller uses the default value of `50000`.

> * If the field has the value of `-1`, then HAProxy dynamically computes a maximum value based on the available `ulimits` in the running container. This process results in a large computed value that incurs significant memory usage compared to the current default value of `50000`.

Author: Yeah it's very irregular everywhere--backticks/no backticks. But I think here with backticks and noun is more consistent with a parameter discussion. Good eyes, thank you for catching these!

* If the field has a value that is greater than the current operating system limit, the `HAProxy` processes do not start.

* If you choose a discrete value and the router pod is migrated to a new node, it is possible that the new node does not have an identical `ulimit` configured. In such cases, the pod fails to start.

> * If you have nodes with different `ulimits` configured, and you choose a discrete value, it is recommended to use the value of `-1` for this field so that the maximum number of connections is calculated at runtime.

Reviewer: ```suggestion
* If you have nodes with different `ulimits` configured, and you choose a discrete value, you can use the value of `-1` for this field so that the maximum number of connections is calculated at runtime.
```
